src/tests/ftest/slurm_setup.py

```diff
@@ -10,279 +10,481 @@
-        "rm -f /var/spool/slurmctld/clustername"]
-
-    if execute_cluster_cmds(all_nodes, cmd_list, args.sudo) > 0:
-        return 1
-
-    # Startup the slurm control service
-    status = execute_cluster_cmds(args.control, SLURMCTLD_STARTUP, args.sudo)
-    if status > 0 or args.debug:
-        execute_cluster_cmds(args.control, SLURMCTLD_STARTUP_DEBUG, args.sudo)
-    if status > 0:
-        return 1
-
-    # Startup the slurm service
-    status = execute_cluster_cmds(all_nodes, SLURMD_STARTUP, args.sudo)
-    if status > 0 or args.debug:
-        execute_cluster_cmds(all_nodes, SLURMD_STARTUP_DEBUG, args.sudo)
-    if status > 0:
-        return 1
-
-    # ensure that the nodes are in the idle state
-    cmd_list = ["scontrol update nodename={} state=idle".format(args.nodes)]
-    status = execute_cluster_cmds(args.nodes, cmd_list, args.sudo)
-    if status > 0 or args.debug:
-        cmd_list = SLURMCTLD_STARTUP_DEBUG
-        execute_cluster_cmds(args.control, cmd_list, args.sudo)
-        cmd_list = SLURMD_STARTUP_DEBUG
-        execute_cluster_cmds(all_nodes, cmd_list, args.sudo)
-    if status > 0:
-        return 1
-    return 0
+class SlurmSetupException(Exception):
+    """Exception for SlurmSetup class."""
+
+
+class SlurmSetup():
+    """Slurm setup class."""
+
+    EPILOG_FILE = '/var/tmp/epilog_soak.sh'
+    EXAMPLE_FILES = [
+        '/etc/slurm/slurm.conf.example',
+        '/etc/slurm/cgroup.conf.example',
+        '/etc/slurm/slurmdbd.conf.example']
+    MUNGE_DIR = '/etc/munge'
+    MUNGE_KEY = '/etc/munge/munge.key'
+    PACKAGE_LIST = ['slurm', 'slurm-example-configs', 'slurm-slurmctld', 'slurm-slurmd']
+    SLURM_CONF = '/etc/slurm/slurm.conf'
+    SLURM_LOG_DIR = '/var/log/slurm'
+
+    def __init__(self, log, nodes, control_node, sudo=False):
+        """Initialize a SlurmSetup object.
+
+        Args:
+            log (logger): object configured to log messages
+            nodes (NodeSet): slurm nodes
+            control_node (NodeSet): slurm control node
+            sudo (bool, optional): whether or not to use sudo with commands. Defaults to False.
+        """
+        self.log = log
+        self.nodes = NodeSet(nodes)
+        self.control = NodeSet(control_node)
+        self.root = 'root' if sudo else None
+
+    @property
+    def all_nodes(self):
+        """Get all the nodes specified in this class.
+
+        Returns:
+            NodeSet: all the nodes specified in this class
+        """
+        return self.nodes.union(self.control)
+
+    def remove(self):
+        """Remove slurm packages from the nodes.
+
+        Raises:
+            SlurmSetupException: if there is a problem removing the packages
+        """
+        self.log.info("Removing slurm packages")
+        result = remove_packages(self.log, self.all_nodes, self.PACKAGE_LIST, self.root)
+        if not result.passed:
+            raise SlurmSetupException("Error removing slurm packages on {result.failed_hosts}")
```

Comment:
Missing string format
```suggestion
            raise SlurmSetupException(f"Error removing slurm packages on {result.failed_hosts}")
```

Reply:
Thanks.